Resource transformations round up milli values before multiplication

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->

**What happened**:

[Resource transformations](https://kueue.sigs.k8s.io/docs/tasks/manage/administer_cluster_quotas/#transform-resources-for-quota-management) round up milli values before multiplication.

With the following config, `300m` CPUs are transformed to `3000` CPU credits, instead of `300`.

```
transformations:
- input: cpu
  outputs:
    example.com/cpu-credits: 1000
  strategy: Retain
```

**What you expected to happen**:

Resource transformations account for decimal quantity values during multiplication.

- `300m * 1000` -> `300`
- `300M * 1m` -> `300k`

**How to reproduce it (as minimally and precisely as possible)**:

https://github.com/wafrelka/kueue/commit/e9a795905ed531528e4bc5d96e2df2e92c84f35c#diff-a1c9ac01d392160cf78127805c49aad2b49ce36fbe1c1a82984cccd9e8c0fa3bR649-R688

Without the fix, the test fails:

```text
--- FAIL: TestNewInfo (0.00s)
    --- FAIL: TestNewInfo/transformMilliValues (0.00s)
        workload_test.go:697: NewInfo(_) = (-want,+got):
              &workload.Info{
                ... // 1 ignored field
                TotalRequests: []workload.PodSetResources{
                        {
                                Name: "",
                                Requests: resources.Requests{
            -                           s"example.com/cpu-credits":    3000,
            +                           s"example.com/cpu-credits":    300,
                                        s"example.com/memory-credits": 300000,
                                },
                                Count:           1,
                                TopologyRequest: nil,
                                ... // 2 identical fields
                        },
                },
                ClusterQueue:   "",
                LastAssignment: nil,
                ... // 1 ignored and 4 identical fields
              }
FAIL
FAIL    sigs.k8s.io/kueue/pkg/workload  0.011s
FAIL
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/go.mod
+++ b/go.mod
@@ -25,6 +25,7 @@ require (
 	go.uber.org/mock v0.6.0
 	go.uber.org/zap v1.28.0
 	golang.org/x/sync v0.22.0
+	gopkg.in/inf.v0 v0.9.1
 	k8s.io/api v0.36.2
 	k8s.io/apiextensions-apiserver v0.36.2
 	k8s.io/apimachinery v0.36.2
@@ -156,7 +157,6 @@ require (
 	google.golang.org/grpc v1.81.1 // indirect
 	google.golang.org/protobuf v1.36.12-0.20260120151049-f2248ac996af // indirect
 	gopkg.in/evanphx/json-patch.v4 v4.13.0 // indirect
-	gopkg.in/inf.v0 v0.9.1 // indirect
 	gopkg.in/natefinch/lumberjack.v2 v2.2.1 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	k8s.io/code-generator v0.36.2 // indirect

--- a/pkg/workload/workload.go
+++ b/pkg/workload/workload.go
@@ -27,9 +27,11 @@ import (
 	"time"
 
 	"github.com/go-logr/logr"
+	"gopkg.in/inf.v0"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
 	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
 	resourcehelpers "k8s.io/component-helpers/resource"
@@ -528,13 +530,12 @@ func applyResourceTransformations(input corev1.ResourceList, transforms map[core
 			// the value of the resource specified in MultiplyBy.
 			if mapping.MultiplyBy != "" {
 				if q, ok := input[mapping.MultiplyBy]; ok {
-					inputQuantity.Mul(q.Value())
+					inputQuantity = multiplyResourceQuantities(inputQuantity, q)
 				}
 			}
 
 			for outputName, baseFactor := range mapping.Outputs {
-				outputQuantity := baseFactor.DeepCopy()
-				outputQuantity.Mul(inputQuantity.Value())
+				outputQuantity := multiplyResourceQuantities(inputQuantity, baseFactor)
 				if accumulated, ok := output[outputName]; ok {
 					outputQuantity.Add(accumulated)
 				}
@@ -548,6 +549,14 @@ func applyResourceTransformations(input corev1.ResourceList, transforms map[core
 		}
 	}
 	return output
+}
+
+func multiplyResourceQuantities(value, mul resource.Quantity) resource.Quantity {
+	value = value.DeepCopy()
+	mul = mul.DeepCopy()
+	product := inf.Dec{}
+	product.Mul(value.AsDec(), mul.AsDec())
+	return *resource.NewDecimalQuantity(product, value.Format)
 }
 
 func CanBePartiallyAdmitted(wl *kueue.Workload) bool {

--- a/pkg/workload/workload_test.go
+++ b/pkg/workload/workload_test.go
@@ -646,6 +646,46 @@ func TestNewInfo(t *testing.T) {
 				},
 			},
 		},
+		"transformMilliValues": {
+			workload: *utiltestingapi.MakeWorkload("transform", "").
+				PodSets(
+					*utiltestingapi.MakePodSet("", 1).
+						Request(corev1.ResourceCPU, "100m").
+						Request(corev1.ResourceMemory, "100M").
+						Obj(),
+				).
+				Obj(),
+			infoOptions: []InfoOption{WithResourceTransformations([]config.ResourceTransformation{
+				{
+					Input:    corev1.ResourceCPU,
+					Strategy: ptr.To(config.Replace),
+					Outputs: corev1.ResourceList{
+						"example.com/cpu-credits": resource.MustParse("3000"),
+					},
+				},
+				{
+					Input:    corev1.ResourceMemory,
+					Strategy: ptr.To(config.Replace),
+					Outputs: corev1.ResourceList{
+						"example.com/memory-credits": resource.MustParse("3m"),
+					},
+				},
+			})},
+			wantInfo: Info{
+				TotalRequests: []PodSetResources{
+					{
+						Name: "",
+						Requests: resources.Requests{
+							// 100m * 3000 = 300
+							corev1.ResourceName("example.com/cpu-credits"): 300,
+							// 100M * 3m = 300k
+							corev1.ResourceName("example.com/memory-credits"): 300 * 1000,
+						},
+						Count: 1,
+					},
+				},
+			},
+		},
 	}
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {

